use kubectl-with-retry on pause & resume

metral opened the pull request:
attempts to fix https://github.com/kubernetes/kubernetes/issues/25645 by using `kubectl-with-retry` on `rollout {pause,resume}` (`resume` is for safe measures) instead of `kubectl` directly, as is done with other `rollout {pause,resume}` tests in this same script.

metral committed:
use kubectl-with-retry on pause & resume

eparis commented:
ok to test

metral commented:
The `e2e` and `Node e2e` tests are hitting the following errors:
```
ERROR: (gcloud.compute.firewall-rules.delete) Some requests did not succeed:
 - The resource 'projects/kubernetes-jenkins-pull/global/firewalls/e2e-gce-agent-pr-3-0-minion-e2e-gce-agent-pr-3-0-http-alt' was not found
ERROR: (gcloud.compute.firewall-rules.delete) Some requests did not succeed:
 - The resource 'projects/kubernetes-jenkins-pull/global/firewalls/e2e-gce-agent-pr-3-0-minion-e2e-gce-agent-pr-3-0-nodeports' was not found
Bringing down cluster using provider: gce
```

```
ERROR: gcloud crashed (OSError): [Errno 39] Directory not empty: '/usr/local/share/google/google-cloud-sdk.staging/lib/googlecloudsdk/third_party/apis'

If you would like to report this issue, please run the following command:
  gcloud feedback
Project: kubernetes-jenkins-pull
Zone: us-central1-f
INSTANCE_GROUPS=
NODE_NAMES=
Bringing down cluster
ERROR: gcloud failed to load: No module named argcomplete
```

```
WARNING: You are creating a legacy network. Using --mode=legacy will be required in future releases.
ERROR: (gcloud.compute.networks.create) Some requests did not succeed:
 - Quota 'ROUTES' exceeded.  Limit: 170.0

2016/05/25 06:28:42 e2e.go:206: Error running up: exit status 1
2016/05/25 06:28:42 e2e.go:202: Step 'up' finished in 47.621947144s
2016/05/25 06:28:42 e2e.go:114: Error starting e2e cluster. Aborting.
```

eparis commented:
@k8s-bot e2e test this flake: #26271

janetkuo commented on hack/test-cmd.sh:
```diff
@@ -1087,13 +1087,11 @@ __EOF__
   kube::test::get_object_assert deployment "{{range.items}}{{$deployment_image_field}}:{{end}}" "${IMAGE_NGINX}:${IMAGE_NGINX}:"
   kube::test::if_has_string "${output_message}" "Object 'Kind' is missing"
   ## Pause the deployment
-  output_message=$(! kubectl rollout pause -f hack/testdata/recursive/deployment --recursive 2>&1 "${kube_flags[@]}")
```
Would you change the title comment of this test too (L1077)? These tests are about "--recursive" and the "rollout" commands, but not about "Rollback a deployment". It's somewhat misleading. Changing each sub-title would help too (maybe add "recursively" to the end of each sub-title, and mention that we expect error message "Object 'Kind' is missing").

metral replied:
Sure thing

janetkuo commented on hack/test-cmd.sh:
```diff
@@ -1087,13 +1087,11 @@ __EOF__
-  output_message=$(! kubectl rollout pause -f hack/testdata/recursive/deployment --recursive 2>&1 "${kube_flags[@]}")
+  kubectl-with-retry rollout pause -f hack/testdata/recursive/deployment --recursive 2>&1 "${kube_flags[@]}"
   kube::test::get_object_assert deployment "{{range.items}}{{.spec.paused}}:{{end}}" "true:true:"
-  kube::test::if_has_string "${output_message}" "Object 'Kind' is missing"
```
I believe we still want this check?

metral replied:
We do, but I removed it as the only solution to triage the flake was to attempt using `kubectl-with-retry` instead of `kubectl`.

However, in doing so, `kubectl-with-retry` directs stderr to a file that is removed once it is done executing, and does not allow me to review it for this missing object as I was doing

janetkuo replied:
We can have a try function with customized error message check, or we can pass the expected error message to that retry function. WDYT?

janetkuo replied:
Do we still need 2>&1 when we're using kubectl-with-retry?

janetkuo replied:
Maybe creating a new retry function with your own check is easier, and we can add a comment saying that removing this function once #20437 is fixed

metral commented:
The e2e test is hitting flake https://github.com/kubernetes/kubernetes/issues/26210

janetkuo commented:
@k8s-bot e2e test this issue: #IGNORE

metral commented:
@janetkuo I've gone ahead and addressed your feedback:

- cleared up test title comments to describe the tests more effectively
- removed 2>&1 from `kubectl-with-retry` as it wasn't needed
- modified `kubectl-with-retry` slightly to allow me to preserve the error file for review/usage in my tests

PTAL

janetkuo commented:
Thanks, looks good now. Please squash and I'll apply the tag.

metral commented:
@janetkuo all set, thanks!
